### PR TITLE
feat: poll grades and generate certificates

### PR DESCRIPTION
**Pull request by arslanashraf7**
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
- [x] Migrations
  - [x] Migration is backwards-compatible with current production code
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested
- [x] Settings
  - [x] New settings are documented and present in `app.json`
  - [x] New settings have reasonable development defaults, if applicable
  - [ ] Opened issue for DevOps regarding necessary configuration changes to deployed environments

#### What are the relevant tickets?
#717 

#### What's this PR do?
- Ports parts of certificates and grades polling-based Certificates from xPRO ([ref](https://github.com/mitodl/mitxpro/blob/master/courses/tasks.py#L26))
- Integrate with edX grades API and fetch grades and save them in MITxOnline
- Generate certificates based on available grades passed & paid status of user
- A cron job to be scheduled to run every 24 hours which would fetch grades of all the eligible course runs and save the grades and create certificates if the criteria meet
- Some refactoring

#### How should this be manually tested?
Below might be good to read before you start testing it and get familiarized with certain criteria.

**1. Eligible Course runs for cron Job:**
- Run end date is in the future or empty
- Since this job will run every 24 hrs and it is possible a course was ended after the last time the job ran, to align with it we have a `CERTIFICATE_CREATION_DELAY_IN_HOURS (Default=24hrs)` - This is the time we use for picking course runs that ended recently e.g. in the last 24 hours.

**2. Eligible Certificates*
1. If a course run has`self_paced=True` we generate the certificates right away when the job runs (For eligible students)
2.  If a course has `self_paced=False` we generate the certificates only if the `end_date` has passed for the run
3. For a User, The certificate generation would be fulfillment of `Point#1/Point#2`, his grade from edX should be `Passed`, and his enrollment should be Paid(`Verified`) (In xPRO there were no free enrollments)

**3. Now to actually test these scenarios:**
1. You need to configure your local edX devstack with MITxOnline.
2. Create a course in edX and link it to MITxOnline.
3. In edX, create grading for the course by visiting (advanced settings in Studio) with passing grade and grade division for Lab/Home work, Mid, Final. (Try to set max grading portion to lab/Homework for easy testing).
4. Go through the course and get your user to pass the course in edX.
5. You can now run `generate_course_certificates` from Django Admin shell and verify the above scenarios for course run/and certificate eligibility
6. Check `CourseRunGrade` and `CourseRunCertificate` models in Admin for the entries

#### Where should the reviewer start?
- Configure edX with MITxOnline locally
- Create a course in edX through Studio, set grading policy, and link the course to MITxOnline

### CODE REVIEW

**Comment by arslanashraf7** [comment]
Marking this blocked by https://github.com/mitodl/mitxonline/pull/740. I believe that PR should merge first but I'll reiterate on that to see if we can still get this PR reviewed and/or merged.

File: courses/api.py
Comment (rhysyngsun) [inline courses/api.py:603]: I'm not sure that this operation should be deleting certificates and I don't see anything about it in the issue Peter wrote, is there a reason you included this behavior?

File: courses/models.py
Comment (rhysyngsun) [inline courses/models.py:395]: Nit: I think this should probably be renamed a bit - a) boolean fields should be prefixed with either `is_` or `has_` b) the certificates being self paced sounds a bit odd, is it rather that the course run is self paced?

File: courses/tasks.py
Comment (rhysyngsun) [inline courses/tasks.py:57]: The body of this task is complex enough that it should be moved into an `api.py` method (or two).

**Review by rhysyngsun — CHANGES_REQUESTED** [review]
Ran short on time today, but here's a partial code review based on a quick skim of the code.

File: courses/api.py
Comment (arslanashraf7) [reply courses/api.py:603]: Most of the things are migrated from xPRO and this has been part of the certificate process there, I also wondered why. And on thing i could think of was that we have an option for resetting grades for a user in edX and in that case if a user gets his grades reset by Course Admins he is going to have `0` grade in which case I suppose we delete the certificate if it has been created already. But we can change this, let me know what you think?

**Comment by arslanashraf7** [comment]
@rhysyngsun Thanks for the review, I've been addressing the partial feedback and it should be done shortly. Just wanted to sync in case you have continued the review, if not, maybe it might be good idea if you could take the look once i address the current feedback.

File: courses/models.py
Comment (arslanashraf7) [reply courses/models.py]: Atleast for now the self paces would only make sense for certificates but you're right if we make it generic it would make more sense for other conditions that we might place based on self paced of a course run. I've changed it to `is_self_paced`

File: main/settings.py
Comment (rhysyngsun) [inline main/settings.py:836]: I'm wondering what your thoughts are on this, I think this would be a good opportunity to make crontab configurations a lot more flexible, what do you think of this PR?

https://github.com/mitodl/ol-django/pull/92

Comment (arslanashraf7) [reply main/settings.py:836]: I think that will surely make it cleaner and easier for defining the crontab jobs. I can review that PR. 

Although i wonder if we should start migrating to that right from this PR or would it be a good idea to create a PR that migrates all our jobs simultaneously to using the the parser from the ol-django.

Comment (rhysyngsun) [reply main/settings.py:836]: That's a good point, probably worth doing as a separate PR also because it'll probably require a PR in `saltops` so it'd be best to just have one PR on each repo instead of a few here and there.

Comment (arslanashraf7) [reply main/settings.py:836]: I agree, In that case, I think you might be able to continue the review on this PR.

**Comment by pdpinch** [comment]
Thinking ahead a little bit, I have a question:

We have some courses in mitxonline production  that have been up for some time now, as audit only. When we turn on payments, and a user pays for a course, will we create a certificate for the user the next time we sync grades? Or is some other action necessary?

**Comment by arslanashraf7** [comment]
> Thinking ahead a little bit, I have a question:
> 
> We have some courses in mitxonline production that have been up for some time now, as audit only. When we turn on payments, and a user pays for a course, will we create a certificate for the user the next time we sync grades? Or is some other action necessary?

@pdpinch So the current implementation is like mentioned below:

**Criteria for picking course runs for this task:**
1. The course runs that have `end_date=None`
2. The runs which have end_date in future
3. The runs that have ended in the last 24 hrs (This hour value is configurable and the default is **24hrs**).

**Criteria to update grade**
1. Since we want to update the grade on regular bases so for all the course runs picked above we update/create the grades for all the enrolled users without any additional conditions.

**Criteria for generating certificates**
1. First thing we now check is if the course run has `is_self_paced=True` or course `end_date` has passed (For non self-paced we want to generate at the end of the course run).
2. If the criteria in above point#1 are met, we check if the user(s) has a `passed` grade and his/her enrollment is paid, then we generate a certificate.

The answer to your question depends upon the above criteria, in addition, we are also creating a command which we should be able to run to generate grades& certificates and the implementation of the command is like "We won't check for the above criteria mentiond for picking course runs and picking the users based on course end_date or self paced, we just pick the run sync the grades and generate the certificates if the user has passed & paid enrollment".

File: courses/api.py
Comment (rhysyngsun) [reply courses/api.py:607]: That makes sense.

**Review by rhysyngsun — COMMENTED** [review]
Code looks good, I'll give this a functional test in the morning.

File: courses/api.py
Comment (rhysyngsun) [inline courses/api.py:645]: I think you should be passing `now` to this so that the logic across this function uses the same value (this is realistically probably not an issue, but consistency is probably good).

Other than that, LGTM so feel free to merge after changing that.

**Review by rhysyngsun — APPROVED** [review]

**Pull request merged**
